homeassistant/components/eufylife_ble/sensor.py

```diff
@@ -0,0 +1,207 @@
+"""Support for EufyLife sensors."""
+from __future__ import annotations
+
+from contextlib import suppress
+from typing import Any
+
+from homeassistant import config_entries
+from homeassistant.components.bluetooth import async_address_present
+from homeassistant.components.sensor import SensorDeviceClass, SensorEntity
+from homeassistant.const import ATTR_UNIT_OF_MEASUREMENT, STATE_UNAVAILABLE, UnitOfMass
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.util.unit_conversion import MassConverter
+from homeassistant.util.unit_system import US_CUSTOMARY_SYSTEM
+
+from .const import DOMAIN, EUFYLIFE_MODEL_TO_NAME
+from .models import EufyLifeData
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: config_entries.ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the EufyLife sensors."""
+    data: EufyLifeData = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [
+            EufyLifeWeightSensorEntity(data),
+            EufyLifeRealTimeWeightSensorEntity(data),
+        ]
+    )
+    if data.client.supports_heart_rate:
+        async_add_entities([EufyLifeHeartRateSensorEntity(data)])
+
+
+class EufyLifeSensorEntity(SensorEntity):
+    """Representation of an EufyLife sensor."""
+
+    _attr_has_entity_name = True
+
+    def __init__(self, data: EufyLifeData) -> None:
+        """Initialize the weight sensor entity."""
+        self._data = data
+
+        self._attr_device_info = DeviceInfo(
+            name=EUFYLIFE_MODEL_TO_NAME[data.model],
+            connections={(dr.CONNECTION_BLUETOOTH, data.address)},
+        )
+
+    @property
+    def available(self) -> bool:
+        """Determine if the entity is available."""
+        if self._data.client.advertisement_data_contains_state:
+            # If the device only uses advertisement data, just check if the address is present.
+            return async_address_present(self.hass, self._data.address)
+
+        # If the device needs an active connection, availability is based on whether it is connected.
+        return self._data.client.is_connected
+
+    @callback
+    def _handle_state_update(self, *args: Any) -> None:
+        """Handle state update."""
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Register callback."""
+        self.async_on_remove(
+            self._data.client.register_callback(self._handle_state_update)
+        )
+
+
+class EufyLifeRealTimeWeightSensorEntity(EufyLifeSensorEntity):
```

Comment:
These classes are very similar.  I'd extend `SensorEntityDescription` instead and add a `value_fn`

Reply:
I thought about doing this, but there is custom logic in some of the entities that I think would make it difficult. For example two of the entities extend RestoreEntity, but the third one doesn't. Even among the two that extend RestoreEntity, the logic is different, since one of them needs to convert the unit and the other one doesn't. I just don't see a way that these three classes could be cleanly combined.

Reply:
Assuming you don't plan on adding any more sensor entities that seems fine.  I'm concerned it could become a bit more difficult to maintain if it turns out to have more sensors.

Reply:
Yeah, I see what you mean. I don't plan on adding any more sensor entities. If I ever do, I'll definitely look into refactoring this to reduce the duplicated code.